### PR TITLE
UI: Add multiple management server support

### PR DESCRIPTION
**Pull request by utchoang**
### Description
Related https://github.com/apache/cloudstack-primate/pull/898
This PR for Add multiple management server support

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### Screenshots (if appropriate):
image

### CODE REVIEW

**Comment by davidjumani** [comment]
Think it would be good to show the management server to which the user is connected listed in the header next to his name. Thoughts @utchoang @rhtyd

**Comment by rhtyd** [comment]
@blueorangutan ui

File: ui/src/components/header/UserMenu.vue [hunk 1]
```diff
@@ -20,6 +20,15 @@
 
     <translation-menu class="action"/>
     <header-notice class="action"/>
+    <label class="user-menu-server-info action">
+      <a-icon
+        slot="prefix"
+        type="database"
+        :style="[{
+          color: $config.theme['@primary-color']
```
Comment (davidjumani) [inline ui/src/components/header/UserMenu.vue:28]: Think it would be alright if just grey

**Comment by davidjumani** [comment]
@utchoang Did you face any CORS issue ?

**Comment by utchoang** [comment]
@davidjumani In the previous discussion at https://github.com/apache/cloudstack-primate/pull/898 there was a discussion about the CORS issue, and I think it will be solved by Nginx proxy. I think so.

**Comment by utchoang** [comment]
@blueorangutan ui

**Comment by wido** [comment]
> @davidjumani In the previous discussion at [apache/cloudstack-primate#898](https://github.com/apache/cloudstack-primate/pull/898) there was a discussion about the CORS issue, and I think it will be solved by Nginx proxy. I think so.

Indeed. As long as a proxy in between is used to use URLs and not different domains for management servers CORS is not an issue, for example:

- /proxy/management-server-1
- /proxy/management-server-2
- /proxy/management-server-3

These would be the relative URLs to reach the different management servers.

**Review by wido — APPROVED** [review]
LGTM

**Comment by weizhouapache** [comment]
I have tested https://github.com/apache/cloudstack-primate/pull/898, which worked.
I used nginx or pfsense/haproxy to solve CORS issue.

@utchoang  are these two prs same ?

**Comment by utchoang** [comment]
@weizhouapache Yes. Two PRs same.

**Review by weizhouapache — APPROVED** [review]
@utchoang great. lgtm

File: ui/src/components/widgets/Console.vue [hunk 2]
```diff
@@ -18,7 +18,7 @@
 <template>
   <a
     v-if="['vm', 'systemvm', 'router', 'ilbvm'].includes($route.meta.name) && 'updateVirtualMachine' in $store.getters.apis"
-    :href="'/client/console?cmd=access&vm=' + resource.id"
+    :href="server + '/client/console?cmd=access&vm=' + resource.id"
```
Comment (davidjumani) [inline ui/src/components/widgets/Console.vue:21]: @utchoang `/client` can be changed in config.json. Would be better to pick it up from apiBase and trim off the /api

**Comment by davidjumani** [comment]
> > @davidjumani In the previous discussion at [apache/cloudstack-primate#898](https://github.com/apache/cloudstack-primate/pull/898) there was a discussion about the CORS issue, and I think it will be solved by Nginx proxy. I think so.
> 
> Indeed. As long as a proxy in between is used to use URLs and not different domains for management servers CORS is not an issue, for example:
> 
>     * /proxy/management-server-1
> 
>     * /proxy/management-server-2
> 
>     * /proxy/management-server-3
> 
> 
> These would be the relative URLs to reach the different management servers.
> 
> This means that you do not set _apiHost_, but only set _apiBase_ for each Management Server.
> 
> Otherwise you need to deal with CORS indeed and thus add headers which can be done by Nginx/Apache.

Thanks, was playing around in dev mode but got it working

**Review by GabrielBrascher — APPROVED** [review]
LGTM

**Comment by nvazquez** [comment]
LGTM
@utchoang there is only one open comment, can you please check?

**Review by davidjumani — APPROVED** [review]

**Comment by DaanHoogland** [comment]
enough lgtm, but has this been tested? (@davidjumani ?)

**Comment by rhtyd** [comment]
@utchoang can you send some doc PR on how users will use it; for example how to configure the reverse proxy/server etc and will the dropdown be hidden for most users, say who don't want to specify or use this feature?

**Comment by davidjumani** [comment]
@DaanHoogland This has been tested and working. For anyone who needs to test it locally, 

Update `ui/public/config.json`
```
  "servers": [
    {
      "name": "QA-Server",
      "apiHost": "",
      "apiBase": "/qa/api"
    },
    {
      "name": "Localhost",
      "apiHost": "",
      "apiBase": "/client/api"
    }
  ],
```

Update `ui/vue.config.js`
```
proxy: {
      '/qa': {
        target: 'http://qa.cloudstack.cloud:8080/',
        secure: false,
        ws: false,
        pathRewrite: {
          '/qa/api': '/client/api',
          '/qa/console': '/console'
        },
        cookiePathRewrite: {
          '/qa/api': '/client/api',
          '*': '/'
        },
        changeOrigin: true,
        proxyTimeout: 10 * 60 * 1000 // 10 minutes
      },
      '/client': {
        target: process.env.CS_URL || 'http://localhost:8080',
        secure: false,
        ws: false,
        changeOrigin: true,
        proxyTimeout: 10 * 60 * 1000 // 10 minutes
      }
    }
```

**Comment by rhtyd** [comment]
@blueorangutan ui

File: ui/public/config.json [hunk 3]
```diff
@@ -1,5 +1,12 @@
 {
   "apiBase": "/client/api",
+  "servers": [
```
Comment (rhtyd) [inline ui/public/config.json:3]: I think this feature shouldn't be enabled by default, wouldn't it confuse users/admins? @utchoang @davidjumani @wido @nvazquez @sureshanaparti @weizhouapache @Pearl1594 @shwstppr ?

Comment (wido) [reply ui/public/config.json:3]: If the list only shows one server at the time, what would it confuse them? There is only one option to select

Comment (utchoang) [reply ui/public/config.json:3]: @rhtyd Yes. I will set it to default hidden mode. If the user/admin wants to enable it, the configuration can be changed.

**Comment by davidjumani** [comment]
@blueorangutan ui

File: ui/public/config.json (hunk 3 above)
Comment (davidjumani) [reply ui/public/config.json:3]: Looks good with @utchoang's change, disabled by default. +1 for getting this in asap

**Review by davidjumani — APPROVED** [review]

File: ui/public/config.json (hunk 3 above)
Comment (nvazquez) [reply ui/public/config.json:3]: +1
thanks @utchoang

**Review by nvazquez — APPROVED** [review]
LGTM

**Comment by nvazquez** [comment]
Merging based on approvals

**Pull request merged**